### PR TITLE
PMIC-PERIODIC

### PR DESCRIPTION
**Pull request by 19deleow**

### CODE REVIEW

**Comment by 19deleow** [comment]
Now this branch is updated

File: CougSat1-PMIC/src/PMICConfiguration.h [hunk 1]
```diff
@@ -149,4 +149,48 @@
...
 #define THERM_CAL_PV3_FRONT_BOTTOM_350K (0.852)
 
+
+// Comparison Constant for Thermistor readings in celcius
+#define TEMP_MIN (0)
+#define TEMP_MAX (50)
+#define REG_TEMP_MAX (85)
+
+// Priority Levels
+#define priority_PR3V3_one      100
```
Comment (WattsUp) [inline CougSat1-PMIC/src/PMICConfiguration.h]: Terrible naming conventions, I can't tell where these are being used. Read the code SOP

File: CougSat1-PMIC/src/PMICConfiguration.h [hunk 2]
```diff
@@ -149,4 +149,48 @@
 #define THERM_CAL_PV3_FRONT_BOTTOM_270K (0.158)
 #define THERM_CAL_PV3_FRONT_BOTTOM_350K (0.852)
 
+
+// Comparison Constant for Thermistor readings in celcius
+#define TEMP_MIN (0)
+#define TEMP_MAX (50)
```
Comment (WattsUp) [inline CougSat1-PMIC/src/PMICConfiguration.h]: Min and max of what? Battery? Also naming conventions:
THRES_BATT_TEMP_LOW (0)
THRES_BATT_TEMP_HIGH (50)
THRES_REG_TEMP_HIGH (85)

File: CougSat1-PMIC/src/components/PowerNode.h [hunk 3]
```diff
@@ -21,20 +21,32 @@
...
+
+  uint8_t getCurrent(double * current);
+
+  void    getSwitch(bool * pathA, bool * pathB);
+
   uint8_t setSwitch(bool pathA, bool pathB);
 
+  uint8_t getVoltage(double * voltage);
+
+  uint8_t getPriority(int priority);
```
Comment (WattsUp) [inline CougSat1-PMIC/src/components/PowerNode.h]: getPriority and getExpectedCurrent is not implemented.

File: CougSat1-PMIC/src/components/PowerNode.h [hunk 4]
```diff
@@ -21,20 +21,32 @@
...
-  LTC2499 &adc;
+  LTC2499 & adc;
   LTC2499Channel_t channel;
-
   double shunt;
-  bool pathA;
-  bool pathB;
+  bool   pathA;
+  bool   pathB;
+  int    priorityN;
```
Comment (WattsUp) [inline CougSat1-PMIC/src/components/PowerNode.h]: Bad variable name, call it priority and expectedCurrent.

File: CougSat1-PMIC/src/components/PowerNodeDirect.h [hunk 5]
```diff
@@ -22,13 +22,15 @@
 #include "drivers/LTC2499.h"
 #include "mbed.h"
 
-class PowerNodeDirect : public PowerNode {
+class PowerNodeDirect : public PowerNode 
+{
 public:
-  PowerNodeDirect(LTC2499 &adc, LTC2499Channel_t channel, double shunt,
+  PowerNodeDirect(LTC2499 & adc, LTC2499Channel_t channel, double shunt,
```
Comment (WattsUp) [inline CougSat1-PMIC/src/components/PowerNodeDirect.h]: These constructors need the priority and expected current added to the constructor. You should have a compilation error currently

File: CougSat1-PMIC/src/events/EventPeriodic.cpp [hunk 6]
```diff
@@ -17,9 +17,629 @@
...
-uint8_t eventPeriodic() { return ERROR_NOT_SUPPORTED; }
\ No newline at end of file
+uint8_t eventPeriodic()
+{
+
+	// From powerrail psuedocode, implement control algorithm
+	// Control input switching to power
+
+	// Monitor battery
+	long BattVoltA = 0; // voltage level of battery A
```
Comment (WattsUp) [inline CougSat1-PMIC/src/events/EventPeriodic.cpp]: Fix variable names

File: CougSat1-PMIC/src/events/EventPeriodic.cpp [hunk 7]
```diff
@@ -17,9 +17,629 @@
...
+		ERROR("Periodic", "Failed to get voltage of Battery B");
+		return result;
+	}
+
+	if (BattVoltA > BattVoltB)
+	{
+		// if voltage of battery A is greater than the voltage of battery of B, redirect current to voltage B
+		// interate through nodesPVOut[] and set each one to go to battery B
+
+		while (nodesPVOut[index] < 8)
```
Comment (WattsUp) [inline CougSat1-PMIC/src/events/EventPeriodic.cpp]: Why not a contained for loop? BTW, this while loop is comparing a pointer to 8

File: CougSat1-PMIC/src/events/EventPeriodic.cpp [hunk 8]
```diff
@@ -17,9 +17,629 @@
...
+	{
+		ERROR("Periodic", "Failed to get current exiting Battery A");
+		return result;
+	}
+
+
+	// monitor current and reduce payload distribution
+	// by turing off payloads that draw current 
+	if (BattCurrA > 4 && BattCurrB > 4)
+	{
```
Comment (WattsUp) [inline CougSat1-PMIC/src/events/EventPeriodic.cpp]: Not sure what you are doing here and variable names and comments are not helping

File: CougSat1-PMIC/src/events/EventPeriodic.cpp [hunk 9]
```diff
@@ -17,9 +17,629 @@
...
+			nodesBatteryHeater[1].setSwitch(false, true);
+		else
+			// both batteries power heater
+			nodeBatteryHeaters[1].setSwitch(true, true);
+	}
+	else
+	{
+		// if temperature reads 10 degrees higher than min temp or if temp reads higher
+		// higher than max temp, turn of heaters.
+		if( temp1 > (TEMP_MIN + 10) || temp1 > TEMP_MAX)
```
Comment (WattsUp) [inline CougSat1-PMIC/src/events/EventPeriodic.cpp]: I like the hysteresis, move the magic number to PMICConfiguration.
If a battery is too hot then its load or charging needs to be reduced

File: CougSat1-PMIC/src/events/EventPeriodic.cpp [hunk 10]
```diff
@@ -17,9 +17,629 @@
...
+	 
+
+	/*******************************************
+	 * Monitor input/output current of both    *
+	 * Regulators. Use getCurrent() to get     *
+	 * current levels of the regulators using  *
+	 * regulator objects. Reduce payloads if   *
+	 * current is above 3A. 
+	 * ****************************************/
+	for(int x = 0; x < 12; x++)
```
Comment (WattsUp) [inline CougSat1-PMIC/src/events/EventPeriodic.cpp]: Once again, not sure what this code does

File: CougSat1-PMIC/src/events/EventPeriodic.cpp (hunk 7 above)
Comment (19deleow) [reply CougSat1-PMIC/src/events/EventPeriodic.cpp]: What do you mean by a contained loop? Like include this section in a single loop.

**Comment by WattsUp** [comment]
Remove your code about balancing the batteries. That will be implemented in hardware. The PMIC's role is just monitoring and turning off loads that are too much

File: CougSat1-PMIC/src/components/PowerNode.h (hunk 4 above)
Comment (WattsUp) [reply CougSat1-PMIC/src/components/PowerNode.h]: expected current should be a double

**Comment by WattsUp** [comment]
Keep this pull request open, don't need to open a new one every commit. Also don't say things like final in commit messages or you will get to final final real final commit

Please address the previous comments and ensure the code compiles. My looking over finds many compiler errors.

File: CISLibrary/components/CurrentNode.cpp [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+/**
+ * @brief Get the aggregate rank (current * rank)
+ *
+ * @param aggregateRank to return to
+ * @return double (current * rank)
+ */
+double CurrentNode::getAggregateRank() {
+  double current = 0.0, aggregateRank;
```
Comment (WattsUp) [inline CISLibrary/components/CurrentNode.cpp]: Update the doxygen. make this function into a 1 liner

File: CISLibrary/components/CurrentNode.cpp [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+/**
+ * @brief Update the current by reading the ADC's value and doing maths
+ * Sets lastCurrent field
+ *
+ * @return mbed_error_status_t
+ */
+mbed_error_status_t CurrentNode::updateCurrent() {
+  return MBED_ERROR_UNSUPPORTED;
```
Comment (WattsUp) [inline CISLibrary/components/CurrentNode.cpp]: Implement this function

File: CougSat1-PMIC/src/events/EventPeriodic.cpp [hunk 13]
```diff
@@ -1,12 +1,247 @@
...
+      int                 count = 0;
+      mbed_error_status_t error = MBED_SUCCESS;
+
+      while (currentToShed > 0.0) 
+      {
+        // Could get stuck in this loop if all nodes are off but there is more
+        // current to shed TODO fix this
+        double        maxAggregate = 0.0;
+        CurrentNode * maxNode      = nullptr;
+
```
Comment (WattsUp) [inline CougSat1-PMIC/src/events/EventPeriodic.cpp]: move uint8_t count here to be reset every loop

File: CougSat1-PMIC/src/events/EventPeriodic.cpp [hunk 14]
```diff
@@ -1,12 +1,247 @@
...
+        // Find the node with the largest aggregate rank that is enabled
+        for (int i = 0; i < nodeCount; ++i)
+        {
+            buf = nodes[i]->getAggregateRank();
+            if (buf > maxAggregate && nodes[i]->getSwitch())
```
Comment (WattsUp) [inline CougSat1-PMIC/src/events/EventPeriodic.cpp]: if (getSwitch()){
++count;
if(buf>maxAggregate){
max = buf
}
}
after the for loop, count is number of enabled nodes. if 0 return.

**Pull request merged**